wsmaster/che-core-api-machine/src/main/java/org/eclipse/che/api/machine/server/MachineService.java

```diff
@@ -251,6 +254,18 @@ public MachineProcessDto executeCommandInMachine(@ApiParam(value = "Machine ID")
                              .collect(Collectors.toList());
     }
 
+    @GET
+    @Path("/{machineId}/recipe")
+    @Produces(MediaType.TEXT_PLAIN)
+    @ApiOperation(value = "Retrieve machine recipe script")
+    @ApiResponses({@ApiResponse(code = 200, message = "Recipe content retrieved"),
+                   @ApiResponse(code = 404, message = "Machine with specified ID does not exist"),
+                   @ApiResponse(code = 500, message = "Internal server error occurred")})
+    public String getRecipeScript(@ApiParam(value = "Machine ID") @PathParam("machineId") String id)
+            throws NotFoundException, ServerException {
+        return recipeRetriever.getRecipe(machineManager.getMachine(id).getConfig()).getScript();
```

Comment:
it will be 500, since it's not a clients error, and it is most appropriate code for server error.
also, added validation of location URL in workspace validator.